### PR TITLE
4.0.0 alpha.1

### PR DESCRIPTION
**Pull request by swilly22**

### CODE REVIEW

File: src/procedures/proc_vector_query.c [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+	SIValue query_vector;  // query vector
+
+	// extract arguments from map
+	if(!_extractArgs(map,
+					 &k,
+					 &et,
+					 &label,
+					 &attribute,
+					 &query_vector)) {
+		//ErrorCtx_SetError("", err);
```
Comment (AviAvni) [inline src/procedures/proc_vector_query.c:253]: handle error message

File: src/serializers/decoders/prev/v11/decode_graph_schema.c [hunk 2]
```diff
@@ -53,8 +62,8 @@ static void _RdbLoadFullTextIndex
 
 	if(!already_loaded) {
 		ASSERT(idx != NULL);
-		Index_SetLanguage(idx, language);
-		Index_SetStopwords(idx, stopwords);
+		if(language  != NULL) Index_SetLanguage(idx, language);
+		if(stopwords != NULL) Index_SetStopwords(idx, &stopwords);
```
Comment (AviAvni) [inline src/serializers/decoders/prev/v11/decode_graph_schema.c:66]: check if this if is needed

File: src/serializers/serializers_include.h [hunk 3]
```diff
@@ -8,27 +8,29 @@
 
 // A single header file including all important headers for serialization.
 
-// Include Redis Modules API.
+// include Redis Modules API
```
Comment (AviAvni) [inline src/serializers/serializers_include.h:11]: remove this coomments

File: tests/flow/constraint_utils.py [hunk 4]
```diff
@@ -85,13 +85,24 @@ def create_mandatory_constraint(g, entity_type, lbl, *props, sync=False):
     return create_constraint(g, "MANDATORY", entity_type, lbl, *props, sync=sync)
 
 def create_unique_node_constraint(g, lbl, *props, sync=False):
-    # create exact-match index
-    create_node_exact_match_index(g, lbl, *props, sync=False)
+    # create exact-match index(es)
+    for p in props:
```
Comment (AviAvni) [inline tests/flow/constraint_utils.py:89]: breaking change

File: tests/flow/test_index_delete.py [hunk 5]
```diff
@@ -1,145 +1,301 @@
...
 
         # validate no indexes in graph
-        result = list_indicies(g)
+        result = list_indicies(self.g)
         self.env.assertEquals(len(result.result_set), 0)
 
-    def test03_drop_exact_match_edge_index(self):
-        # create an index over edge follow:created_at
-        result = create_edge_exact_match_index(g, 'follow', 'created_at', sync=True)
+    def test_06_index_rollback(self):
```
Comment (AviAvni) [inline tests/flow/test_index_delete.py:186]: move to test_undo_log.py

File: tests/flow/test_index_delete.py [hunk 6]
```diff
@@ -1,145 +1,301 @@
...
-                plan = g.execution_plan(q)
+                plan = self.g.execution_plan(q)
                 self.env.assertNotIn("Index", plan)
 
-                indicies = list_indicies(g).result_set
+                indicies = list_indicies(self.g).result_set
                 self.env.assertEquals(len(indicies), 0)
 
-    def test06_reset_order(self):
+    def test_08_reset_order(self):
```
Comment (AviAvni) [inline tests/flow/test_index_delete.py:275]: move this test

**Review by AviAvni — CHANGES_REQUESTED** [review]

File: src/index/index.c [hunk 7]
```diff
@@ -116,49 +195,44 @@ void Index_ConstructStructure
 	RediSearch_IndexOptionsSetGCPolicy(idx_options, GC_POLICY_FORK);
 	#endif
 
+	RediSearch_IndexOptionsSetStopwords(idx_options, NULL, 0);
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/index/index.c:198]: it is.
```
127.0.0.1:6379> GRAPH.QUERY g "CREATE FULLTEXT INDEX FOR (n:N) ON (n.a) OPTIONS {language:'enz'}"
1) 1) "Indices created: 1"
   2) "Cached execution: 0"
   3) "Query internal execution time: 0.365791 milliseconds"
127.0.0.1:6379> GRAPH.QUERY g "CREATE FULLTEXT INDEX FOR (n:N) ON (n.b) OPTIONS {language:'dadad'}"
(error) Can not override index configuration
127.0.0.1:6379>
```

File: src/index/index.c [hunk 8]
```diff
@@ -168,127 +242,118 @@ RSDoc *Index_IndexGraphEntity
...
-			for(uint i = 1; i < none_indexable_fields_count; i++) {
-				len += sprintf(s + len, "%c%s", INDEX_SEPARATOR, none_indexable_fields[i]);
-			}
+		if(field->type & INDEX_FLD_VECTOR && (t & T_VECTOR)) {
+			*doc_field_count += 1;
 
-			RediSearch_DocumentAddFieldString(doc, INDEX_FIELD_NONE_INDEXED,
-						s, len, RSFLDTYPE_TAG);
+			size_t   n        = SIVector_ElementsByteSize(*v);
+			uint32_t dim      = SIVector_Dim(*v);
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/index/index.c:306]: We crash, fixed. dim mis-matched vectors are skipped (not indexed)

File: tests/flow/test_index_delete.py (hunk 6 above)
Comment (swilly22) [reply tests/flow/test_index_delete.py:275]: Couldn't find an alternative test file for this test, the only reason for this test to remain here is related to the fact that it relies on the existence of an index.

File: tests/flow/constraint_utils.py (hunk 4 above)
Comment (swilly22) [reply tests/flow/constraint_utils.py:89]: Indeed this is a breaking change, I think this is the right path to take here.

File: src/serializers/decoders/prev/v11/decode_graph_schema.c (hunk 2 above)
Comment (swilly22) [reply src/serializers/decoders/prev/v11/decode_graph_schema.c:66]: yes, this is required, if there are no `stopwords` then stopwords will be `NULL` and we don't want to pass NULL to `Index_SetStopwords`

Comment (swilly22) [reply src/serializers/decoders/prev/v11/decode_graph_schema.c:66]: Although I believe we can remove the condition for `language`

**Pull request merged**